tests/simplex/periodic_tet.cc

```diff
@@ -0,0 +1,300 @@
+ * governing the source code and code contributions can be found in
+ * LICENSE.md and CONTRIBUTING.md at the top level directory of deal.II.
+ *
+ * ------------------------------------------------------------------------
+ *
+ * Test that the function
+ * GridGenerator::subdivided_hyper_rectable_with_simplices() can give a mesh
+ * which is suitible for periodicity with tetrahedra.
+ */
+
+#include <deal.II/base/types.h>
+
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_tools.h>
+
+#include <deal.II/fe/fe_simplex_p.h>
+#include <deal.II/fe/mapping_fe.h>
+
+#include <deal.II/grid/filtered_iterator.h>
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_out.h>
+#include <deal.II/grid/grid_tools.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/tria_accessor.h>
+#include <deal.II/grid/tria_iterator.h>
+#include <deal.II/grid/tria_orientation.h>
+
+#include <deal.II/lac/affine_constraints.h>
+
+#include "../tests.h"
+
+void
+check()
+{
+  Triangulation<3>          tria;
+  Point<3>                  corner_p1     = {-0.5, 1.0, 0.0};
+  Point<3>                  corner_p2     = {0.5, 2.0, 1.2};
+  std::vector<unsigned int> sub_divisions = {2, 2, 2};
+  GridGenerator::subdivided_hyper_rectangle_with_simplices(
+    tria, sub_divisions, corner_p1, corner_p2, true, true);
+
+  // Check that we have made a grid with the expected number of cells and only
+  // one level of refinement.
+  if (tria.n_cells() != 48)
+    {
+      deallog << "FAIL! Wrong number of cells." << std::endl;
+    }
+
+  if (tria.n_levels() != 1)
+    {
+      deallog << "FAIL! Mesh has more than one refinement level." << std::endl;
+    }
+
+  // Check that each cell has a positive measure.
+  {
+    for (const auto &cell : tria.active_cell_iterators())
+      {
+        double measure = cell->measure();
+        if (measure <= 0.0)
+          {
+            deallog << "FAIL! Found a cell with nonpositive measure: "
+                    << cell->active_cell_index() << std::endl;
+          }
+      }
+  }
+
+  // Collect periodic face pairs and ensure that they make sense and we have
+  // the right numbers of them.
+  std::array<std::vector<GridTools::PeriodicFacePair<
+               typename Triangulation<3>::cell_iterator>>,
+             3>
+    pairs;
+  GridTools::collect_periodic_faces(tria, 0, 1, 0, pairs[0]);
+  GridTools::collect_periodic_faces(tria, 2, 3, 1, pairs[1]);
+  GridTools::collect_periodic_faces(tria, 4, 5, 2, pairs[2]);
+  tria.add_periodicity(pairs[0]);
+  tria.add_periodicity(pairs[1]);
+  tria.add_periodicity(pairs[2]);
+
+  // To enable periodicity the easy way, we need each boundary pair to be in
+  // default orientation. Check that this holds.
+  {
+    IteratorFilters::AtBoundary filter;
+    const unsigned char         default_orientation =
+      ReferenceCells::Tetrahedron.default_combined_face_orientation();
+    const unsigned char default_flip_orientation =
```

Comment:
'flip' means a 120 degree rotation for triangles - could you use the world 'mirror' here instead?